storybookjs/storybook
# Svelte: Improve CSF3 types

## What I did

This PR provides improved type safety for Svelte stories (similar to what we changed to React) but requires svelte-check instead of tsc (Svelte for VS Code extension for editor support):
https://github.com/sveltejs/language-tools/tree/master/packages/svelte-check

Requires the satisfies keyword available in TS ≥ 4.9:
https://github.com/microsoft/TypeScript/pull/46827

Built on top of this CSF PR:
https://github.com/ComponentDriven/csf/pull/51

### Typesafe args
We changed `StoryObj` and `Meta` to increase type safety for when the user provides args partially in meta.

Considering a Component like this:

```svelte
<script lang="ts">
  import { createEventDispatcher } from 'svelte';

  export let disabled: boolean;
  export let label: string;

  const dispatch = createEventDispatcher();
</script>

<button on:click={(e) => dispatch('clicked', e)} on:dblclick on:mousemove {disabled}>
  {label}
</button>
```

It is valid to provide args like this:

```ts
import Button from './Button.svelte';

// ✅ valid
const meta = {
  component: Button,
  args: { disabled: false },
} satisfies Meta<Button>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Basic = {
  args: { label: 'good' }
} satisfies Story;
```

While it is invalid to forget an arg, in either meta or the story:

```ts
// ❌ invalid
const meta = {
  component: Button,
} satisfies Meta<Button>;

export const Basic = {
  args: { label: 'good' }
} satisfies Story;

// ❌ invalid
const meta = {
  component: Button,
  args: { label: 'good' }
} satisfies Meta<Button>;

export const Basic = {} satisfies Story;
```

Changed Meta to make sure both a Component, as the Props of the component can be used:

```ts
const meta = {
  component: Button,
} satisfies Meta<Button>;

export default meta;

export const meta = {
  component: Button,
} satisfies Meta<ComponentProps<Button>>;

export default meta;
```

If the component is used, you will have enhanced autocompletion for events (maybe in a future version also for slots, when we support that):

```ts
const meta = {
  component: Button,
  args: { label: 'good', disabled: false },
  render: (args) => ({
    Component: Button,
    props: args,
    on: {
      mousemove: (event) => {
        expectTypeOf(event).toEqualTypeOf<CustomEvent>();
      },
    },
  }),
} satisfies Meta<{ disabled: boolean; label: string }>;
```

### Typesafe decorators/loaders

Decorators now accept a new generic arg argument to be specified:

```ts
const withDecorator: DecoratorFn<{ decoratorArg: string }> = (
  storyFn,
  { args: { decoratorArg } }
) => ({
  Component: Decorator,
  props: { decoratorArg },
});
```

And the type of meta/story will check if this arg is part of the generic type:

```ts
type Props = ComponentProps<Button> & { decoratorArg: string };

const meta = satisfies<Meta<Props>>()({
  component: Button,
  args: { disabled: false },
  decorators: [withDecorator],
});

type Story = StoryObj<typeof meta>;
const Basic: Story = { args: { decoratorArg: 0, label: 'good' } };
```

## How to test

You can test it by running: `yarn nx run @storybook/svelte:check`

@benmccann @j3rem1e If you could have a look :)
